src/Components/Components/test/NavigationManagerTest.cs

```diff
@@ -208,6 +215,356 @@ public void GetUriWithQueryParameters_ThrowsWhenAnyParameterNameIsEmpty(string b
         Assert.StartsWith("Cannot have empty query parameter names.", exception.Message);
     }
 
+    [Fact]
+    public void LocationChangingHandlers_CanContinueTheNavigationSynchronously_WhenOneHandlerIsRegistered()
+    {
+        // Arrange
+        var baseUri = "scheme://host/";
+        var navigationManager = new TestNavigationManager(baseUri);
+        navigationManager.AddLocationChangingHandler(HandleLocationChanging);
+
+        // Act
+        var navigation1 = navigationManager.RunNotifyfLocationChangingAsync($"{baseUri}/subdir1", null, false);
+
+        // Assert
+        Assert.True(navigation1.IsCompletedSuccessfully);
+        Assert.True(navigation1.Result);
+
+        static ValueTask HandleLocationChanging(LocationChangingContext context)
+        {
+            return ValueTask.CompletedTask;
+        };
+    }
+
+    [Fact]
+    public void LocationChangingHandlers_CanContinueTheNavigationSynchronously_WhenMultipleHandlersAreRegistered()
+    {
+        // Arrange
+        var baseUri = "scheme://host/";
+        var navigationManager = new TestNavigationManager(baseUri);
+        var initialHandlerCount = 3;
+        var completedHandlerCount = 0;
+
+        // Act
+        for (var i = 0; i < initialHandlerCount; i++)
+        {
+            navigationManager.AddLocationChangingHandler(HandleLocationChanging);
+        }
+
+        var navigation1 = navigationManager.RunNotifyfLocationChangingAsync($"{baseUri}/subdir1", null, false);
+
+        // Assert
+        Assert.True(navigation1.IsCompletedSuccessfully);
+        Assert.True(navigation1.Result);
+        Assert.Equal(initialHandlerCount, completedHandlerCount);
+
+        ValueTask HandleLocationChanging(LocationChangingContext context)
+        {
+            completedHandlerCount++;
+            return ValueTask.CompletedTask;
+        };
+    }
+
+    [Fact]
+    public async Task LocationChangingHandlers_CanContinueTheNavigationAsynchronously_WhenOneHandlerIsRegistered()
+    {
+        // Arrange
+        var baseUri = "scheme://host/";
+        var navigationManager = new TestNavigationManager(baseUri);
+        navigationManager.AddLocationChangingHandler(HandleLocationChanging);
+
+        // Act
+        var navigation1 = navigationManager.RunNotifyfLocationChangingAsync($"{baseUri}/subdir1", null, false);
+        var navigation1Result = await navigation1.WaitAsync(Timeout);
+
+        // Assert
+        Assert.True(navigation1.IsCompletedSuccessfully);
+        Assert.True(navigation1Result);
+
+        static async ValueTask HandleLocationChanging(LocationChangingContext context)
+        {
+            await Task.Delay(1000);
```

Comment:
This will introduce flakyness in the build.

You can use a TaskCompletionSource instead to deterministically decide when to complete the handler task. (See other tests in Renderer if you want examples)

Reply:
I think what you want to do here is similar to what we do in the Renderer tests. Use a TCS on the handler, call
`var navigation1 = navigationManager.RunNotifyfLocationChangingAsync($"{baseUri}/subdir1", null, false);`

Assert that it hasn't completed (as you would expect). Set the result on the TCS and assert that it completed and produced the result you expected. 

You shouldn't need to have the Timeout